pkg/controllers/webhookcerts/csr_approver.go

```diff
@@ -0,0 +1,130 @@
+package webhookcerts
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	certapiv1 "k8s.io/api/certificates/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	certv1informers "k8s.io/client-go/informers/certificates/v1"
+	certv1client "k8s.io/client-go/kubernetes/typed/certificates/v1"
+	certv1listers "k8s.io/client-go/listers/certificates/v1"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+type csrApproverController struct {
+	csrClient certv1client.CertificateSigningRequestInterface
+	csrLister certv1listers.CertificateSigningRequestLister
+}
+
+// saGroupSet is the set of groups for the SA expected to have created the CSR
+var saGroupSet = sets.NewString(
+	"system:serviceaccounts",
+	"system:serviceaccounts:openshift-authentication-operator",
+	"system:authenticated",
+)
+
+// NewCSRApproverController returns a controller that is observing the CSR API
+// for a CSR named "system:openshift:openshift-authenticator".
+// If such a CSR exists, it checks that it was submitted by the system:serviceaccount:openshift-authentication-operator:authentication-operator
+// SA and if so and it hasn't yet been approved, it auto-approves it
+func NewCSRApproverController(
+	operatorClient v1helpers.OperatorClient,
+	csrClient certv1client.CertificateSigningRequestInterface,
+	csrInformers certv1informers.CertificateSigningRequestInformer,
+	eventsRecorder events.Recorder,
+) factory.Controller {
+	c := &csrApproverController{
+		csrClient: csrClient,
+		csrLister: csrInformers.Lister(),
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithSyncDegradedOnError(operatorClient).
+		WithFilteredEventsInformers(nameFilter(csrName), csrInformers.Informer()).
+		ToController("WebhookAuthenticatorCertApprover", eventsRecorder.WithComponentSuffix("webhook-authenticator-cert-approver"))
+}
+
+func (c *csrApproverController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	csr, err := c.csrLister.Get(csrName)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	if csr.Spec.Username != "system:serviceaccount:openshift-authentication-operator:authentication-operator" {
+		return fmt.Errorf("CSR %q was created by an unexpected user: %q", csrName, csr.Spec.Username)
+	}
+
+	if csrGroups := sets.NewString(csr.Spec.Groups...); !csrGroups.IsSuperset(saGroupSet) {
+		return fmt.Errorf("CSR %q was created by a user with unexpected groups: %v", csrName, csrGroups.List())
```

Comment:
what is the result of these? We go degraded? Or just silence?

Reply:
the `WithSyncDegradedOnError(operatorClient).` line in the creation of this controller makes sure errors in the sync cause degraded state